File: crates/taflow-core/src/indicators/high_low_index.rs
use crate::error::{TaError, TaResult};
use std::collections::VecDeque;
/// Simple mean of the percentage of new extremes that are new highs.
#[derive(Debug, Clone)]
pub struct HighLowIndex {
    period: usize,
    rows: VecDeque<f64>,
    sum: f64,
    count: usize,
    value: Option<f64>,
}
impl HighLowIndex {
    /// Create a rolling state with a positive smoothing period.
    pub fn new(period: usize) -> TaResult<Self> {
        if period == 0 {
            return Err(TaError::InvalidParameter {
                name: "period",
                value: period.to_string(),
                reason: "must be positive",
            });
        }
        Ok(Self {
            period,
            rows: VecDeque::with_capacity(period),
            sum: 0.0,
            count: 0,
            value: None,
        })
    }
    /// Append aggregate new-high and new-low counts for one market tick.
    pub fn append(&mut self, new_highs: f64, new_lows: f64) -> Option<f64> {
        self.count += 1;
        let percent = 100.0 * new_highs / (new_highs + new_lows).max(1.0);
        if self.rows.len() == self.period {
            self.sum -= self.rows.pop_front().expect("full window");
        }
        self.rows.push_back(percent);
        self.sum += percent;
        self.value = (self.rows.len() == self.period).then(|| self.sum / self.period as f64);
        self.value
    }
    /// Return the number of processed market ticks.
    pub fn len(&self) -> usize {
        self.count
    }
    /// Return the latest smoothed percentage, or `None` during warm-up.
    pub fn value(&self) -> Option<f64> {
        self.value
    }
    /// Clear the rolling window while retaining its allocation.
    pub fn reset(&mut self) {
        self.rows.clear();
        self.sum = 0.0;
        self.count = 0;
        self.value = None;
    }
}

File: python/taflow/indicators/high_low_index.py
"""Python adapter for the native High-Low Index."""

from typing import Any
import numpy as np
from .._native import HighLowIndex as _Native
from .._series import as_float64_series


class HighLowIndex:
    """Smooth the percentage of new extremes that are new highs.

    Each tick contributes ``100 * new_highs / max(new_highs + new_lows, 1)``;
    native Rust returns its simple mean over ``period`` ticks. Warm-up is
    represented by ``NaN``. This maps to Wickra ``HighLowIndex``.

    Args:
        new_highs: Number of constituents making a new high at each tick.
        new_lows: Number making a new low at each aligned tick.
        period: Smoothing window. Defaults to 10.

    Raises:
        ValueError: If histories differ in length or ``period`` is zero.
    """

    def __init__(self, period: int = 10) -> None:
        """Initialize an empty configured native state.
        """
        self._state = _Native(period)

    def append(self, new_highs: float, new_lows: float) -> "HighLowIndex":
        """Append one aggregate-extremes tick and return this adapter."""
        self._state.append(float(new_highs), float(new_lows))
        return self

    def extend(self, new_highs: Any, new_lows: Any) -> "HighLowIndex":
        """Append aligned extreme-count histories after length validation."""
        arrays = as_float64_series(new_highs), as_float64_series(new_lows)
        if len(arrays[0]) != len(arrays[1]):
            raise ValueError("new-high and new-low counts must have equal lengths")
        self._state.extend(*arrays)
        return self

    @property
    def value(self) -> float | None:
        """Return the latest smoothed percentage, or ``None`` during warm-up."""
        return self._state.value

    def compute(self) -> np.ndarray:
        """Return aligned High-Low Index history with warm-up as ``NaN``."""
        return self._state.compute()

    def reset(self) -> "HighLowIndex":
        """Clear the rolling window and return this adapter."""
        self._state.reset()
        return self

    def __len__(self) -> int:
        """Return the processed-tick count delegated to native state."""
        return len(self._state)
